### DISCUSSION TITLE
Cannot reduce number of segments

### DISCUSSION THREAD

**Discussion by gland1** [milvus-io/milvus#33847] [Q&A and General discussion]
We are using 50M wiki  data set of 768 d vectors
We have measured diskann ~30000 QPS/0.9 recall on a single graph
With milvus we reach only 200-300 QPS.
We try to reduce the number of segments by increasing :
       dataCoord:
          segment:
            maxSize: 81920
            diskSegmentMaxSize: 81920
            sealProportion: 0.9
            smallProportion: 0.5

But I can only acheive 42 segments.
Why cant we reach a smaller number of segments?
I expected that by increasing the  max segment size and seal proportion, to reduce the number of
segments to about 4-8
Tried manual compaction did not help reducing the number of segments

**Comment by gland1** [comment]
I don't know if it helps but I can see errors during compaction, it seems that datanode gets timeouts writing to minio.. I don't see why it should happen since minio is up and for example query nodes are able to load data from it.. maybe compaction
tries to put a very large file hence the timeout

<html>
<body>

024/06/13 13:03:21.348 +00:00] [WARN] [storage/remote_chunk_manager.go:152] ["failed to put object"] [bucket=milvus3] [path=files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697] [error="IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697]: context deadline exceeded"]
--
Thu, Jun 13 2024 4:03:21 pm | [2024/06/13 13:03:21.348 +00:00] [WARN] [retry/retry.go:46] ["retry func failed"] [traceID=40abf13427c7079fce0ba70390742cb5] [retried=0] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697]: context deadline exceeded"]
Thu, Jun 13 2024 4:03:21 pm | [2024/06/13 13:03:21.350 +00:00] [WARN] [datanode/compactor.go:393] ["failed to upload single insert log"] [planID=450433074137518527] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697]: context deadline exceeded"]
Thu, Jun 13 2024 4:03:21 pm | [2024/06/13 13:03:21.350 +00:00] [WARN] [datanode/compactor.go:544] ["compact wrong, fail to merge"] [traceID=40abf13427c7079fce0ba70390742cb5] [planID=450433074137518527] ["timeout in seconds"=900] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697]: context deadline exceeded"]
Thu, Jun 13 2024 4:03:21 pm | [2024/06/13 13:03:21.351 +00:00] [WARN] [datanode/compaction_executor.go:113] ["compaction task failed"] [planID=450433074137518527] [Collection=450420049857702595] [channel=milvus3-rootcoord-dml_1_450420049857702595v1] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/101/450433074136708698]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/0/450433074136708699\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136707375/1/450433074136708700\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136707375/100/450433074136708697]: context deadline exceeded"]
Thu, Jun 13 2024 4:16:21 pm | [2024/06/13 13:16:21.394 +00:00] [WARN] [storage/remote_chunk_manager.go:152] ["failed to put object"] [bucket=milvus3] [path=files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139] [error="IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139]: context deadline exceeded"]
Thu, Jun 13 2024 4:16:21 pm | [2024/06/13 13:16:21.394 +00:00] [WARN] [storage/remote_chunk_manager.go:152] ["failed to put object"] [bucket=milvus3] [path=files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140] [error="IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140]: context deadline exceeded"]
Thu, Jun 13 2024 4:16:21 pm | [2024/06/13 13:16:21.394 +00:00] [WARN] [storage/remote_chunk_manager.go:152] ["failed to put object"] [bucket=milvus3] [path=files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141] [error="IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141]: context deadline exceeded"]
Thu, Jun 13 2024 4:16:21 pm | [2024/06/13 13:16:21.394 +00:00] [WARN] [retry/retry.go:46] ["retry func failed"] [traceID=798b15478675f9c7cc84ae50baaf2888] [retried=0] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141]: context deadline exceeded"]
Thu, Jun 13 2024 4:16:21 pm | [2024/06/13 13:16:21.396 +00:00] [WARN] [datanode/compactor.go:393] ["failed to upload single insert log"] [planID=450433074137522818] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141]: context deadline exceeded"]
Thu, Jun 13 2024 4:16:21 pm | [2024/06/13 13:16:21.396 +00:00] [WARN] [datanode/compactor.go:544] ["compact wrong, fail to merge"] [traceID=798b15478675f9c7cc84ae50baaf2888] [planID=450433074137522818] ["timeout in seconds"=900] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141]: context deadline exceeded"]
Thu, Jun 13 2024 4:16:21 pm | [2024/06/13 13:16:21.396 +00:00] [WARN] [datanode/compaction_executor.go:113] ["compaction task failed"] [planID=450433074137522818] [Collection=450420049857702595] [channel=milvus3-rootcoord-dml_0_450420049857702595v0] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/101/450433074136709139]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/0/450433074136709140]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708660/1/450433074136709141]: context deadline exceeded"]
Thu, Jun 13 2024 4:19:21 pm | [2024/06/13 13:19:21.370 +00:00] [WARN] [storage/remote_chunk_manager.go:152] ["failed to put object"] [bucket=milvus3] [path=files/insert_log/450420049857702595/450420049857702596/450433074136708709/101/450433074136709460] [error="IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/101/450433074136709460]: context deadline exceeded"]
Thu, Jun 13 2024 4:19:21 pm | [2024/06/13 13:19:21.371 +00:00] [WARN] [storage/remote_chunk_manager.go:152] ["failed to put object"] [bucket=milvus3] [path=files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461] [error="IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461\": context deadline exceeded"]
Thu, Jun 13 2024 4:19:21 pm | [2024/06/13 13:19:21.371 +00:00] [WARN] [storage/remote_chunk_manager.go:152] ["failed to put object"] [bucket=milvus3] [path=files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462] [error="IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462\": context deadline exceeded"]
Thu, Jun 13 2024 4:19:21 pm | [2024/06/13 13:19:21.371 +00:00] [WARN] [retry/retry.go:46] ["retry func failed"] [traceID=53b421187bc3e5a13062e6821701cfd6] [retried=0] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708709/101/450433074136709460: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/101/450433074136709460]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462\": context deadline exceeded"]
Thu, Jun 13 2024 4:19:21 pm | [2024/06/13 13:19:21.372 +00:00] [WARN] [datanode/compactor.go:393] ["failed to upload single insert log"] [planID=450433074137523722] [error="failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708709/101/450433074136709460: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/101/450433074136709460]: context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136708709/0/450433074136709461\": context deadline exceeded: failed to write files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462: IO failed[key=files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462]: Put \"http://milvus3-minio.kioxia:9000/milvus3/files/insert_log/450420049857702595/450420049857702596/450433074136708709/1/450433074136709462\": context deadline exceeded"]


</body>
</html>

**Comment by yhmo** [comment]
You are using v2.4.
There is a timeout setting for compaction: https://github.com/milvus-io/milvus/blob/15b78749a6fdc5258bf3d3824ea13477bc7c1f76/internal/datanode/compactor.go#L453

The default timeout value for compaction is 900 seconds: https://github.com/milvus-io/milvus/blob/15b78749a6fdc5258bf3d3824ea13477bc7c1f76/pkg/util/paramtable/component_param.go#L2882

You set the segment.maxSize = 80GB
A compaction task reads some segments from S3, compacts them in memory, and uploads the large segment to S3. If the task cannot be finished in 900 seconds, you will get the "context deadline exceeded" error.

Enlarge the default value `dataCoord.compaction.timeout` in milvus.yaml. Or use a smaller `segment.maxSize` value.

**Comment by gland1** [comment]
thanks for your help .., I've actually noticed this param(dataCoord.compaction.timeout) last evening and tried ran it for the night and today
I see I have 3 flushed segments and 15 dropped.
but when I loaded the collection , One query node got 15 segments and one in for each of the other 2 query nodes.
Naturally, latency was much larger in the qn with 15 segments.
I had a storage issue in between so I'm trying to figure what's going on

**Comment by gland1** [comment]
what exactly are the dropped segments ? I thought they should not be loaded to qn

**Comment by gland1** [comment]
![image](https://github.com/milvus-io/milvus/assets/50872480/275c2099-23ab-4335-a5f5-47624c4e8429)
see , during load, it was more or less balanced 6/6/5 but it ended like that

**Comment by gland1** [comment]
after a while , the dropped segments disappeared I now only have 3 flushed segments and can reach 4600 QPS.
It goes to show how important is segment sizing and how difficult it is to get it right.
I think dynamic sizing would have been more suitable and user friendly -
The problem is that milvus doesn't know in advance the collection size - but it can get an estimation from the user during collection creation.. the size the segment according to number of query nodes and index size..

**Reply by yhmo** [reply]
We don't suppose users to understand every detail of the internal machinery of milvus. It is a complicate system and difficult to explain.
As we know, milvus manages data by "segments", when you continually insert data into milvus, the data is accumulated in "growing segment". Once the size of growing segment exceed a threshold, a growing segment will be turned into "sealed segment". Sealed segment is static, cannot be change any more. The threshold is configured in the milvus.yaml: https://github.com/milvus-io/milvus/blob/f67b6dc2b054bd5c5a057e73b108cf1c4889d0c8/configs/milvus.yaml#L428

segment.maxSize * segment.sealProportion is the threshold.
Let's say you insert data batch by batch continually. There will be small sealed segments generated with size = 100MB+
Compaction machinery is to merge these small segments into large segments.
Let's say there are 2 small segment A and B.
If you have called collection.load(). the A and B will be kept in query node's memory for search.
Now A and B is in memory.
Compaction begins to merge A and B to C.
After C generated, the index node will build index for C.
After index of C is ready, the query node will load the C_index into memory, and evict A and B from memory.
Now C replace A and B. A and B will be marked as "soft-delete", waiting the GC process to permanently delete them.
So you will see there are 2 "dropped" segment.
GC process triggers every hour. Eventually A and B are deleted from storage.

So, you will see numbers of segments in the monitoring system, growing, sealed, dropped.
Only the growing segments and sealed segments are loaded in memory.

What is size of segment? If a segment is not indexed, the original data of the segment is loaded into memory, the size is the original data size. If the segment is indexed, the query node will load its index into memory to take place the original data, then the size is the size of index. 
It is a dynamic process, segment has different stage, the memory usage is not const if the data is continually changed. The tool estimates the final state of the memory usage, but it cannot estimate the states in different stage. If data insertion is stop, the dynamic process will continue for even hours. During the dynamic process, you will see the number of segments changed, the memory usage changed, the disk usage changed, the search performance changed, etc...

**Comment by gland1** [comment]
thanks for your help and info

**Comment by gland1** [comment]
does diskAnn index performance scale horizontally?
I mean, I agree that for example if an index has 3 segments, qps will be higher on 3 nodes than in one.
but if there's only one segment, will single node with one segment for the same index achive a better result than than the 3 nodes with 3 segments?  I think it should from diskann tests we've conducted.
can you a suggest a test scenario which will demonstrate horizontal scaling with diskAnn for large datasets?

**Reply by xiaofan-luan** [reply]
> does diskAnn index performance scale horizontally? I mean, I agree that for example if an index has 3 segments, qps will be higher on 3 nodes than in one. but if there's only one segment, will single node with one segment for the same index achive a better result than than the 3 nodes with 3 segments? I think it should from diskann tests we've conducted. can you a suggest a test scenario which will demonstrate horizontal scaling with diskAnn for large datasets?

The truth is, you can  not build segment size with unlimited size, reason is:
1. index build time is much longer
2. update and delete will cause index rebuild
3. load segment takes more extra memory
4. load can not be balanced between nodes.

So in milvus we limit each segment to be less than a threshold, typically 4-8G at most.

But you are right, increase segment size will improve performance in general

**Comment by xiaofan-luan** [comment]
we strongly don't recommend to tune the segment size to any value larger than 10GB.
you will see many strange behaviors like timeout. 
My guess is build a 80GB diskann index will take more than one day. 
With 50M data, we've been working on distributed index for quite well. it will be on zilliz cloud first and later for opensource

**Reply by gland1** [reply]
When I create the index when the collection is empty - total insert time with 100000 vectors iterations takes about 6 hours.
Then a few more hours for compaction.
Can you share more info about the distributed index you've mentioned?
Is it going to use cluster+graph techniques or other 2 phase search schemes ?

**Reply by xiaofan-luan** [reply]
did you mean 100000 vector as batch size？
you can do bulk insertion to acceletate.

 distributed index means we build another layer on top of current index on each segments and filtering out some segments that far away from query vectors
